src/Workspaces/Core/Portable/Rename/RenameSymbolAnnotation.cs

```diff
@@ -0,0 +1,107 @@
+    /// in some cases to notify the workspace host of refactorings.
+    /// </summary>
+    /// <remarks>
+    /// This annotation is applied to a symbol that has been renamed. 
+    /// When Workspace.TryApplyChanges happens in Visual Studio, we raise rename events for that symbol.
+    /// </remarks>
+    internal static class RenameSymbolAnnotation
+    {
+        public const string RenameSymbolKind = nameof(RenameSymbolAnnotation);
+
+        public static bool ShouldAnnotateSymbol(ISymbol symbol)
+            => symbol.DeclaringSyntaxReferences.Any();
+
+        public static SyntaxAnnotation? Create(ISymbol oldSymbol)
+        {
+            return ShouldAnnotateSymbol(oldSymbol)
+                ? new SyntaxAnnotation(RenameSymbolKind, SerializeData(oldSymbol))
+                : null;
+        }
+
+        public static TNode WithRenameSymbolAnnotation<TNode>(this TNode node, SemanticModel semanticModel)
+            where TNode : SyntaxNode
+            => node.WithAdditionalAnnotations(Create(semanticModel.GetDeclaredSymbol(node)));
+
+        public static async Task<ImmutableDictionary<ISymbol, ISymbol>> GatherChangedSymbolsInDocumentsAsync(
+            IEnumerable<DocumentId> changedDocumentIds,
+            Solution newSolution, Solution oldSolution,
+            CancellationToken cancellationToken)
+        {
+            var changedSymbols = ImmutableDictionary.CreateBuilder<ISymbol, ISymbol>();
+
+            foreach (var changedDocId in changedDocumentIds)
+            {
+                var newDocument = newSolution.GetDocument(changedDocId);
+
+                // Documents without syntax tree won't have the annotations attached
+                if (newDocument is null || !(newDocument.SupportsSyntaxTree))
+                {
+                    continue;
+                }
+
+                var syntaxRoot = await newDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var symbolRenameNodes = syntaxRoot.GetAnnotatedNodes(RenameSymbolKind);
+
+                foreach (var node in symbolRenameNodes)
+                {
+                    foreach (var annotation in node.GetAnnotations(RenameSymbolKind))
+                    {
+                        var oldDocument = oldSolution.GetDocument(changedDocId);
+                        if (oldDocument is null)
+                        {
+                            continue;
+                        }
+
+                        var newSemanticModel = await newDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                        var currentSemanticModel = await oldDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                        var oldSymbol = ResolveSymbol(annotation, currentSemanticModel.Compilation);
+                        var newSymbol = newSemanticModel.GetDeclaredSymbol(node);
```

Comment:
GetDeclaredSymbol could return null, why did it let you pass that to changedSymbols.Add?

Reply:
It _can_, but shouldn't in this case right? Assuming we're always annotating something that declares a symbol (which is the useful point of this). Maybe just add a contract throw?

Reply:
I'm more worried why the compiler didn't yell at you. Is this a bug in nullable somewhere?

Reply:
hmm... the extension method is annotated as return a non-null value. Potentially by design but not desired... 

`SemanticModel.GetDeclaredSymbolForNode` might be annotated wrong?